File: packages/examples/todo/backends/rust/src/commands/clear.rs
use super::{ok, warning, Command};
use crate::store::TodoStore;
use afd::{CommandResult, WarningSeverity};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Clear(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {
    all: Option<bool>,
}

#[derive(Serialize)]
struct Output {
    cleared: usize,
    remaining: usize,
}

impl Command for Clear {
    type Input = Input;
    const NAME: &'static str = "todo-clear";
    const DESCRIPTION: &'static str = "Clear completed todos (or all if specified)";
    const MUTATION: bool = true;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": {
                "all": {
                    "type": "boolean",
                    "description": "If true, clear all todos regardless of status"
                }
            }
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        if input.all == Some(true) {
            let cleared = self.0.clear_all();
            let output = Output {
                cleared,
                remaining: 0,
            };
            return ok(
                &output,
                format!("Cleared all {cleared} todos"),
                1.0,
                Vec::new(),
            );
        }

        let (cleared, remaining) = self.0.clear_completed();
        let (reasoning, warnings) = if cleared > 0 {
            (
                format!("Cleared {cleared} completed todo(s), {remaining} remaining"),
                vec![warning(
                    "PERMANENT",
                    "This action cannot be undone".to_string(),
                    WarningSeverity::Info,
                )],
            )
        } else {
            (
                format!("No completed todos to clear, {remaining} remaining"),
                Vec::new(),
            )
        };
        ok(&Output { cleared, remaining }, reasoning, 1.0, warnings)
    }
}

File: packages/examples/todo/backends/typescript/src/commands/clear.ts
/**
 * @fileoverview todo.clear command
 */

import { z } from "zod";
import { defineCommand, success } from "@lushly-dev/afd-server";
import { store } from "../store/index.js";

const inputSchema = z.object({
  all: z
    .boolean()
    .optional()
    .describe("If true, clear all todos regardless of status"),
});

export interface ClearResult {
  cleared: number;
  remaining: number;
}

export const clearCompleted = defineCommand<typeof inputSchema, ClearResult>({
  name: 'todo-clear',
  description: "Clear completed todos (or all if specified)",
  category: "todo",
  tags: ['todo', 'clear', 'write', 'batch', 'destructive'],
  mutation: true,
  version: "1.0.0",
  input: inputSchema,

  async handler(input: z.infer<typeof inputSchema>) {
    console.error(`[todo.clear] input:`, JSON.stringify(input));
    if (input.all) {
      const count = store.count();
      console.error(`[todo.clear] Clearing all ${count} todos`);
      store.clear();
      return success(
        { cleared: count, remaining: 0 },
        {
          reasoning: `Cleared all ${count} todos`,
          confidence: 1.0,
        }
      );
    }

    const result = store.clearCompleted();

    return success(
      {
        cleared: result.cleared,
        remaining: result.remaining,
      },
      {
        reasoning:
          result.cleared > 0
            ? `Cleared ${result.cleared} completed todo${
                result.cleared === 1 ? "" : "s"
              }, ${result.remaining} remaining`
            : `No completed todos to clear, ${result.remaining} remaining`,
        confidence: 1.0,
        warnings:
          result.cleared > 0
            ? [
                {
                  code: "PERMANENT",
                  message: "This action cannot be undone",
                  severity: "info",
                },
              ]
            : undefined,
      }
    );
  },
});
